zquery/shared/src/main/scala/zio/zquery/Continue.scala

```diff
@@ -0,0 +1,140 @@
+  ): Continue[R1, E1, B] =
+    self match {
+      case Get(io)       => effect(ZQuery.fromEffect(io).foldCauseM(failure, success))
+      case Effect(query) => effect(query.foldCauseM(failure, success))
+    }
+
+  /**
+   * Purely maps over the success type of this continuation.
+   */
+  def map[B](f: A => B): Continue[R, E, B] =
+    self match {
+      case Get(io)       => get(io.map(f))
+      case Effect(query) => effect(query.map(f))
+    }
+
+  /**
+   * Purely maps over the failure type of this continuation.
+   */
+  final def mapError[E1](f: E => E1)(implicit ev: CanFail[E]): Continue[R, E1, A] =
+    self match {
+      case Get(io)       => get(io.mapError(f))
+      case Effect(query) => effect(query.mapError(f))
+    }
+
+  /**
+   * Effectually maps over the success type of this continuation.
+   */
+  def mapM[R1 <: R, E1 >: E, B](f: A => ZQuery[R1, E1, B]): Continue[R1, E1, B] =
+    self match {
+      case Get(io)       => effect(ZQuery.fromEffect(io).flatMap(f))
+      case Effect(query) => effect(query.flatMap(f))
+    }
+
+  /**
+   * Purely contramaps over the environment type of this continuation.
+   */
+  def provideSome[R0](f: Described[R0 => R])(implicit ev: NeedsEnv[R]): Continue[R0, E, A] =
+    self match {
+      case Get(io)       => get(io)
+      case Effect(query) => effect(query.provideSome(f))
+    }
+
+  /**
+   * Runs this continuation..
+   */
+  def runCache(cache: Cache): ZIO[R, E, A] =
+    self match {
+      case Get(io)       => io
+      case Effect(query) => query.runCache(cache)
+    }
+
+  /**
+   * Combines this continuation with that continuation using the specified
+   * function, in sequence.
+   */
+  def zipWith[R1 <: R, E1 >: E, B, C](that: Continue[R1, E1, B])(f: (A, B) => C): Continue[R1, E1, C] =
+    (self, that) match {
+      case (Get(l), Get(r))       => get(l.zipWith(r)(f))
+      case (Get(l), Effect(r))    => effect(ZQuery.fromEffect(l).zipWith(r)(f))
+      case (Effect(l), Get(r))    => effect(l.zipWith(ZQuery.fromEffect(r))(f))
+      case (Effect(l), Effect(r)) => effect(l.zipWith(r)(f))
+    }
+
+  /**
+   * Combines this continuation with that continuation using the specified
+   * function, in parallel.
+   */
+  def zipWithPar[R1 <: R, E1 >: E, B, C](that: Continue[R1, E1, B])(f: (A, B) => C): Continue[R1, E1, C] =
+    (self, that) match {
+      case (Get(l), Get(r))       => get(l.zipWithPar(r)(f))
+      case (Get(l), Effect(r))    => effect(ZQuery.fromEffect(l).zipWithPar(r)(f))
+      case (Effect(l), Get(r))    => effect(l.zipWithPar(ZQuery.fromEffect(r))(f))
+      case (Effect(l), Effect(r)) => effect(l.zipWithPar(r)(f))
+    }
+
+}
+
+private[zquery] object Continue {
+
+  /**
+   * Constructs a continuation from a request, a data source, and a `Ref` that
+   * will contain the result of the request when it is executed.
+   */
+  def apply[R, E, A, B](request: A, dataSource: DataSource[R, A], ref: Ref[Option[Either[E, B]]])(
```

Comment:
I'm not sure if it's useful or not, but you could rewrite that `Ref` to be `Promise` and delete the `Option`, pushing the `Either` into the failure channel of the Promise—if it's only going to be completed once. `Promise[E, B]` is easier to use / understand than `Ref[Option[Either[E, B]]]`.

Reply:
Yes. We use this a bunch of places internally. I thought it was a little more efficient to use `Ref` since we never suspend and it isn't visible to users but I think we could change throughout.